webrender/src/device/gfx/render_pass.rs

```diff
@@ -2,147 +2,145 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use api::ImageFormat;
 use hal::Device;
+use hal::format::Format;
+use hal::image::Layout;
+use hal::pass::{Attachment, SubpassDesc};
+use internal_types::FastHashMap;
 
-pub(super) struct HalRenderPasses<B: hal::Backend> {
-    pub(super) r8: B::RenderPass,
-    pub(super) r8_depth: B::RenderPass,
-    pub(super) bgra8: B::RenderPass,
-    pub(super) bgra8_depth: B::RenderPass,
+pub(super) const DEPTH_FORMAT: Format = hal::format::Format::D32Sfloat;
+
+const SUBPASS_R8: SubpassDesc = SubpassDesc {
+    colors: &[(0, Layout::ColorAttachmentOptimal)],
+    depth_stencil: None,
+    inputs: &[],
+    resolves: &[],
+    preserves: &[],
+};
+
+const SUBPASS_DEPTH_R8: SubpassDesc = SubpassDesc {
+    colors: &[(0, Layout::ColorAttachmentOptimal)],
+    depth_stencil: Some(&(1, Layout::DepthStencilAttachmentOptimal)),
+    inputs: &[],
+    resolves: &[],
+    preserves: &[],
+};
+
+const SUBPASS_BGRA8: SubpassDesc = SubpassDesc {
+    colors: &[(0, Layout::ColorAttachmentOptimal)],
+    depth_stencil: None,
+    inputs: &[],
+    resolves: &[],
+    preserves: &[],
+};
+
+const SUBPASS_DEPTH_BGRA8: SubpassDesc = SubpassDesc {
+    colors: &[(0, Layout::ColorAttachmentOptimal)],
+    depth_stencil: Some(&(1, Layout::DepthStencilAttachmentOptimal)),
+    inputs: &[],
+    resolves: &[],
+    preserves: &[],
+};
+
+pub(super) const DEPTH_ATTACHMENT_STATE: AttachmentState = AttachmentState {
+    format: DEPTH_FORMAT,
+    src_layout: Layout::DepthStencilAttachmentOptimal,
+    dst_layout: Layout::DepthStencilAttachmentOptimal,
+    load_op: hal::pass::AttachmentLoadOp::Load,
+};
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub(super) struct AttachmentState {
+    pub format: Format,
+    pub src_layout: Layout,
```

Comment:
We don't have to put the layout in here. Instead, we can always create the render passes with, say `ColorAttachmentOptimal .. ShaderReadOnlyOptimal` and then do the transitions outside (using `pipeline_barrier`) if we need a different layout at the end, or we want to start with a different layout. That would allow us to have much fewer render passes in the end, not to mention the compatibility concerns (to be explained)